cmd/caa-log-checker/main.go

```diff
@@ -0,0 +1,176 @@
+package main
+
+import (
+	"bufio"
+	"compress/gzip"
+	"crypto/sha256"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/cmd"
+)
+
+func openFile(path string) (*bufio.Scanner, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	var reader io.Reader
+	reader = f
+	if strings.HasSuffix(path, ".gz") {
+		reader, err = gzip.NewReader(f)
+		if err != nil {
+			return nil, err
+		}
+	}
+	scanner := bufio.NewScanner(reader)
+	return scanner, nil
+}
+
+type issuanceEvent struct {
+	SerialNumber string
+	Names        []string
+	ResponseTime time.Time
+	Requester    int64
+}
+
+var raIssuanceLineRE = regexp.MustCompile(`Certificate request - successful JSON=(.*)`)
+
+func checkIssuances(scanner *bufio.Scanner, checkedMap map[string][]time.Time, stderr *os.File) error {
+	lNum := 0
+	for scanner.Scan() {
+		lNum++
+		line := scanner.Text()
+		matches := raIssuanceLineRE.FindStringSubmatch(line)
+		if matches == nil {
+			continue
+		}
+		if len(matches) != 2 {
+			return fmt.Errorf("line %d: unexpected number of regex matches", lNum)
+		}
+		var ie issuanceEvent
+		err := json.Unmarshal([]byte(matches[1]), &ie)
+		if err != nil {
+			return fmt.Errorf("line %d: failed to unmarshal JSON: %s", lNum, err)
+		}
+		var badNames []string
+		for _, name := range ie.Names {
+			nameOk := false
+			for _, t := range checkedMap[name] {
+				if t.Before(ie.ResponseTime) && t.After(ie.ResponseTime.Add(-8*time.Hour)) {
+					nameOk = true
+				}
+			}
+			if !nameOk {
+				badNames = append(badNames, name)
+			}
+		}
+		if len(badNames) > 0 {
+			hash := hashNames(ie.Names)
```

Comment:
I think we don't need hashNames - I was using it when debugging to de-dupe certificates, but for the output of this tool it's not necessary. That also lets us get rid of uniqueLowerNames.